pkg/runner/test_runner_route.go

```diff
@@ -45,109 +43,40 @@ func tryClose(ch chan struct{}) {
-		injectionWg     sync.WaitGroup
-		routeInProgress bool
-	)
-
-	// terminateTargetWriter is a control channel used to signal termination to
-	// the writer object which injects a target into the test step
-	terminateTargetWriter := make(chan struct{})
-
-	// `targets` is used to buffer targets coming from the previous routing blocks,
-	// queueing them for injection into the TestStep. The list is accessed
-	// synchronously by a single goroutine.
-	targets := list.New()
+	log.Debugf("Initializing routeIn for %s", stepLabel)
+	targetWriter := newTargetWriter(log, r.timeouts)
 
 	// `ingressTarget` is used to keep track of ingress times of a target into a test step
 	ingressTarget := make(map[string]time.Time)
-
-	// Channel that the injection goroutine uses to communicate back to `routeIn` the results
-	// of asynchronous injection
-	injectResultCh := make(chan injectionResult)
-
-	// injectionChannels are used to inject targets into test step and return results to `routeIn`
-	injectionChannels := injectionCh{stepIn: r.routingChannels.stepIn, resultCh: injectResultCh}
-
-	log.Debugf("initializing routeIn for %s", stepLabel)
-	targetWriter := newTargetWriter(log, r.timeouts)
-
 	for {
 		select {
 		case <-terminate:
-			err = fmt.Errorf("termination requested for routing into %s", stepLabel)
-		case injectionResult := <-injectResultCh:
-			log.Debugf("received injection result for %v", injectionResult.target)
-			routeInProgress = false
-			if injectionResult.err != nil {
-				err = fmt.Errorf("routing failed while injecting target %+v into %s", injectionResult.target, stepLabel)
-				targetInErrEv := testevent.Data{EventName: target.EventTargetInErr, Target: injectionResult.target}
-				if err := r.ev.Emit(targetInErrEv); err != nil {
-					log.Warningf("could not emit %v event for target %+v: %v", targetInErrEv, *injectionResult.target, err)
-				}
-			} else {
-				targetInEv := testevent.Data{EventName: target.EventTargetIn, Target: injectionResult.target}
-				if err := r.ev.Emit(targetInEv); err != nil {
-					log.Warningf("could not emit %v event for Target: %+v", targetInEv, *injectionResult.target)
-				}
-			}
+			return 0, fmt.Errorf("termination requested for routing into %s", stepLabel)
 		case t, chanIsOpen := <-r.routingChannels.routeIn:
 			if !chanIsOpen {
-				log.Debugf("routing input channel closed")
-				r.routingChannels.routeIn = nil
-			} else {
-				log.Debugf("received target %v in input", t)
-				targets.PushFront(t)
-			}
-		}
-
-		if err != nil {
-			break
-		}
-
-		if routeInProgress {
-			continue
-		}
-
-		// no targets currently being injected in the test step
-		if targets.Len() == 0 {
-			if r.routingChannels.routeIn == nil {
-				log.Debugf("input channel is closed and no more targets are available, closing step input channel")
 				close(r.routingChannels.stepIn)
-				break
+				return len(ingressTarget), nil
+			}
+			log.Debugf("received target %v in input", t)
+			ingressTarget[t.ID] = time.Now()
```

Comment:
Is there any chance that targets will be repeated?
If not it could be just targetsCount++